src/Extension.Tests/ChallengeTests.cs

```diff
@@ -138,5 +138,56 @@ public async Task can_use_on_code_submitted_handler_to_access_evaluations_from_s
                 }
             });
         }
+
+        [Fact]
+        public async Task can_use_rule_handler_to_access_code()
+        {
+            var capturedCode = new List<string>();
+            var lesson = new Lesson();
+            using var kernel = new CompositeKernel
+            {
+                new CSharpKernel()
+            }.UseLessonEvaluateMiddleware(lesson);
+            var challenge = GetEmptyChallenge(lesson);
+            await lesson.StartChallengeAsync(challenge);
+            challenge.AddRule(context =>
+            {
+                capturedCode.Add(context.SubmittedCode);
+            });
+
+            await kernel.SubmitCodeAsync("1 + 1");
+            await kernel.SubmitCodeAsync("1 + 2");
+            await kernel.SubmitCodeAsync("1 + 3");
+
+            capturedCode.Should().BeEquivalentTo("1 + 1", "1 + 2", "1 + 3");
+        }
+
+        [Fact]
+        public async Task can_use_rule_handler_to_access_events()
```

Comment:
This tests would "read" easier if it was "teacher can access events when evaluating rules / challenges"

Reply:
updated